src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceService.ChangeNamespaceCodeAction.cs

```diff
@@ -0,0 +1,526 @@
+                // 1. Change the namespace declaration, fix references and add imports that might be necessary.
+                // 2. Explicitly merge the diff to get a new solution.
+                // 3. Remove added imports that are unnecessary.
+                // 4. Do another explicit diff merge based on last merged solution.
+                //
+                // The reason for doing explicit diff merge twice is so merging after remove unnecessaty imports can be correctly handled.
+
+                var solutionAfterNamespaceChange = _state.Solution;
+                var referenceDocuments = PooledHashSet<DocumentId>.GetInstance();
+
+                foreach (var id in _state.DocumentIds)
+                {
+                    var result = await ChangeNamespaceToMatchFoldersAsync(solutionAfterNamespaceChange, id, cancellationToken).ConfigureAwait(false);
+                    solutionAfterNamespaceChange = result.Item1;
+                    referenceDocuments.AddRange(result.Item2);
+                }
+
+                var solutionAfterFirstMerge = await MergeDiffAsync(_state.Solution, solutionAfterNamespaceChange, cancellationToken).ConfigureAwait(false);
+
+                // After changing documents, we still need to remove unnecessary imports related to our change.
+                // We don't try to remove all imports that might become unnecessary/invalid after the namespace change, 
+                // just ones that fully matche the old or new namespace.
+                // For example, if we are changing namespace `Foo.Bar` to `A.B`, the using of name `Bar` below would remain 
+                // untouched:
+                //
+                //      namespace Foo
+                //      {
+                //          using Bar;
+                //          ~~~~~~~~~
+                //      }
+                //
+                // Also, because we may have added different imports to document that triggered the refactoring
+                // and the documents that reference affected types declared in changed namespace, we try to remove
+                // unnecessary imports separately.
+                var solutionAfterImportsRemoved = await RemoveUnnecessaryImportsAsync(
+                    solutionAfterFirstMerge,
+                    _state.DocumentIds, 
+                    CreateAllContainingNamespaces(OldNamespaceParts), 
+                    cancellationToken).ConfigureAwait(false);
+
+                solutionAfterImportsRemoved = await RemoveUnnecessaryImportsAsync(
+                    solutionAfterImportsRemoved, 
+                    referenceDocuments.ToImmutableArray(), 
+                    ImmutableArray.Create(_state.DeclaredNamespace, _state.TargetNamespace), 
+                    cancellationToken).ConfigureAwait(false);
+
+                var solutionAfterSecondMerge = await MergeDiffAsync(solutionAfterFirstMerge, solutionAfterImportsRemoved, cancellationToken).ConfigureAwait(false);
```

Comment:
i'm really curious what a MergeDiff is...   you only got solutionAfterImportsRemoved by transforming solutionAfterFirstMerge.  So it's unclear to me what it means to then merge the final solution you have with the original solutoin started with.

Reply:
If I don't do a second merge diff explicitly, then when applying the code action, the diff will be merged based on the original solution, instead of the solution after first explicit merge, which will cause problem for multi-targeting projects.